### ISSUE TITLE
Biomemap all zeros on mapblocks above ground level

### ISSUE THREAD

**Issue by CBugDCoder** [minetest/minetest#9384]
##### Minetest version
```
Minetest 5.1.1
```

##### OS / Hardware

Operating system: Kubuntu 19.10 x64
CPU: Intel Pentium Dual-Core T4400

##### Summary

I am attempting to use the biomemap in a lua mapgen but it is returning all zeros above ground. I am generating terrain high in the sky so I have been forced to use minetest.get_biome_data(pos) which does return a proper biome id but makes the generation take twice as long.  

##### Steps to reproduce
Create a simple mod with thin in init.lua
minetest.register_on_generated(function(minp, maxp, seed)
	local bvals = minetest.get_mapgen_object("biomemap")
	print(dump(bvals))
	print(dump(minp))
end)

Then fly around observing the biomemap output and the minp.

**Comment by CBugDCoder** [comment]
After some more debugging I have some more definitive data. At ground level (The mapblock that contains y=0) there are no zeros in the biomemap. Above ground level there are all zeros in the biomemap. And deep below ground level there are some zeros but it's never entire. There are two numbers of zeros that it keeps bouncing between (2304 and 1280) making me think that this is not just some random glitch but it is actually an underlying problem.

**Comment by paramat** [comment]
Yes, this occurs because there is an optimisation that only calculates biome if a non-air node exists in a column. Of course then the biomemap is uncalculated for all-air mapchunks.
This was discussed on the forum recently, so i appreciate an issue being opened for it.

In core mapgen, the biomemap is 2D and contains the biome of the highest solid surface in each column.
For an empty mapchunk there is no way to know which y value to calculate biome for in a column. The position you want to calculate for depends on what a mod generates in that mapchunk. The biomemap calculation code runs before any mod, so it cannot know what y values to use.

Essentially, even if the biomemap had values for an empty mapchunk, they would be values for a certain fixed y (bottom or middle?), so of no use for a mod that needs biome for a specific y.
So this seems impossible to solve, you are much better off calculating biome for a 3D position using the API.

**Comment by CBugDCoder** [comment]
Something like minetest.get_biomes(minp,maxp) would be sufficient.
It would be basically
`
local biomes = []
for z = minp.z,maxp.z do
    for y = minp.y,maxp.y do
        for x = minp.x,maxp.x do
            table.insert(biomes, minetest.get_biome_data({x=x,y=y,z=z}).biome)
        end
    end
end
`
but would run much faster since it will be in C++ and wont have to translate between languages too many times.

**Comment by paramat** [comment]
Yes, we could perhaps have a new API that calculates a 3D biomemap using bulk perlin noise.

**Comment by paramat** [comment]
Hmmm but there is a problem with that.

Mods very rarely need biome calculated for every node in a mapchunk (512000 nodes).
For example floating islands, you would calculate biome for the top surface in a column and use that for the surface and all solid nodes below. Maybe calculate for another possible 'top surface' that may be below. Usually you would only need 1-2 calculations per column, not 80. So the vast majority of the 512000 calculations would be wasted calculations.
However because we do not know which nodes biome is needed for, only a 3D biomemap covers all possible needs.
It is possible that calculating for all 512000 nodes would be slower than, or not much faster than, using the existing API for the 6400 or so nodes you need biome for.

`minetest.get_biome_data(pos)` can be used anytime, not just in 'register on generated', so it has to calculate noise values.
We could possibly have new APIs (whatever they are) that are only usable in 'register on generated' (which is fine for Lua mapgen), which would then use the already-calculated heat and humidity maps, improving performance.

**Comment by CBugDCoder** [comment]
True. You could easily just grab a y-slice by running minetest.get_biomes(minp, {x=maxp.x, y = minp.y, z = maxp.z}). Also would it be possible to use the pre-generated heat and humidity maps when in a mapgen thread and calculate them when not.

**Comment by CBugDCoder** [comment]
And you could possibly have it take an array of positions similar to minetest.bulk_set_node.

**Comment by CBugDCoder** [comment]
I built a workaround using the heat and humidity maps in lua and it runs well.

**Comment by paramat** [comment]
If we add a new API only usable in 'register on generated' i think the best thing to do is return a 2D biomemap for a specified y, defaulting to mapchunk middle (or base). Users mostly only need a biomemap for a particular y or maybe a very few non-consecutive y values.
Always calculating all 512000 nodes would be overkill, intensive, slow and usually unneeded.
